### ISSUE TITLE
Autograd API to get saved for backwards tensors for an autograd graph

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#74222]
### 🐛 Describe the bug

@Chillee nerdsniped me into thinking about how to get AOTAutograd working on higher order derivatives. For a moment, let's forget about the actual compilation aspect, and just focus on what it would take to trace out and then reconstruct a torch.autograd.Function that supports higher order derivatives.

The basic hazard, as laid out in @zdevito's  https://colab.research.google.com/drive/1VpeE6UvEPRz9HmsHh1KS0XxXjYu533EC?usp=sharing is that saved tensors for backwards may receive gradient contributions when run in a higher order context. If we trace only a single backwards, we won't ever see such contributions (because they're always zero in this case.)

So what ingredients do we need for our torch.autograd.Function?

1. We need to explicitly return every saved tensor for backwards as an output of the torch.autograd.Function, so that they are able to receive gradients when used, and
2. We need a way to specify additional gradients for the saved tensors when we run backwards

If we have (1), we automatically have (2), since we can add the list of tensors to the outputs argument of torch.autograd.grad, and then just specify the extra gradient contributions we wanted to provide with `grad_outputs`. So all we need is something like `torch.autograd.saved_for_backwards(outputs, inputs)` which returns a list of all tensors which were saved for backwards given the graph specified by outputs and inputs, so that we can additionally return them as outputs.

Then, AOTAutograd can simply perform this query, and add the saved tensors for backwards as extra outputs to the forward function they trace. Graph partitioning is then trivial, as the cut points are exactly uses of these saved tensors.

### Versions

master

**Comment by albanD** [comment]
Here are two simple ways to do this today if you need it:
```python
import torch

x = torch.rand(2, requires_grad=True)

# Method 1: hook
saved_tensors = []
def hook(x):
    saved_tensors.append(x)
    return x

with torch.autograd.graph.saved_tensors_hooks(hook, lambda x: x):
    y = x + 2
    z = y * x
    out = z.exp()

# The Tensors in the array are the exact same as the one we saved
expected = [y, x, out]
print(tuple(a is b for a, b in zip(saved_tensors, expected)))
# prints (True, True, True)

# Method 2: graph traversal
y = x + 2
z = y * x
out = z.exp()

saved_tensors = []
def find_saved(node):
    if not node:
        return
    # Find the saved Tensors on this node
    for field in dir(node):
        if field.startswith("_saved_"):
            val = getattr(node, field)
            if isinstance(val, torch.Tensor):
                saved_tensors.append(val)
    # Recurse
    for child, _ in node.next_functions:
        find_saved(child)

find_saved(out.grad_fn)

# The Tensors in the array are the exact same as the one we saved
# The order is DIFFERENT from the previous one. This one search from
# the "end" while the other one follows the forward order
expected = [out, x, y]
print(tuple(a is b for a, b in zip(saved_tensors, expected)))
# prints (False, True, True)

# Also because sometimes the saved Tensor is "reconstructed", then
# it is not the same object even though they are equal
print(torch.equal(saved_tensors[0], out))
# prints True


```
